KotshiPageJsonAdapter.java:25: error: illegal start of expression

Error Info:

> KotshiPageJsonAdapter.java:25: error: illegal start of expression
>     adapter0 = moshi.adapter(Types.newParameterizedType(List.class, ? extends T.class))

**adapter0 = moshi.adapter(Types.newParameterizedType(List.class, ? extends T.class));**

VO.kt
```java
@JsonSerializable
data class Page<T> constructor(@Json(name = "my_list") var list: List<T>)

@KotshiJsonAdapterFactory
object ApplicationJsonAdapterFactory : KotshiApplicationJsonAdapterFactory()
```

Generated KotshiPageJsonAdapter.java
```
public final class KotshiPageJsonAdapter<T> extends NamedJsonAdapter<Page<T>> {
  private static final JsonReader.Options OPTIONS = JsonReader.Options.of(
      "my_list");

  private final JsonAdapter<List<? extends T>> adapter0;

  public KotshiPageJsonAdapter(Moshi moshi, Type[] types) {
    super("KotshiJsonAdapter(Page)");
    adapter0 = moshi.adapter(Types.newParameterizedType(List.class, ? extends T.class));
  }

  @Override
  public void toJson(JsonWriter writer, Page<T> value) throws IOException {
    if (value == null) {
      writer.nullValue();
      return;
    }
    writer.beginObject();

    writer.name("my_list");
    adapter0.toJson(writer, value.getList());

    writer.endObject();
  }

  @Override
  public Page<T> fromJson(JsonReader reader) throws IOException {
    if (reader.peek() == JsonReader.Token.NULL) {
      return reader.nextNull();
    }

    reader.beginObject();

    List<? extends T> list = null;
    while (reader.hasNext()) {
      switch (reader.selectName(OPTIONS)) {
        case 0: {
          list = adapter0.fromJson(reader);
          continue;
        }
        case -1: {
          reader.nextName();
          reader.skipValue();
          continue;
        }
      }
    }

    reader.endObject();
    StringBuilder stringBuilder = null;
    if (list == null) {
      stringBuilder = KotshiUtils.appendNullableError(stringBuilder, "list");
    }
    if (stringBuilder != null) {
      throw new NullPointerException(stringBuilder.toString());
    }

    return new Page<T>(
        list);
  }
}

```

--- a/compiler/src/main/kotlin/se/ansman/kotshi/AdapterKey.kt
+++ b/compiler/src/main/kotlin/se/ansman/kotshi/AdapterKey.kt
@@ -1,6 +1,10 @@
 package se.ansman.kotshi
 
-import com.squareup.javapoet.*
+import com.squareup.javapoet.CodeBlock
+import com.squareup.javapoet.ParameterizedTypeName
+import com.squareup.javapoet.TypeName
+import com.squareup.javapoet.TypeVariableName
+import com.squareup.javapoet.WildcardTypeName
 import com.squareup.moshi.Types
 import javax.lang.model.element.Element
 
@@ -32,6 +36,11 @@ data class AdapterKey(
                             }
                             .add(")")
                             .build()
+                is WildcardTypeName -> when {
+                    lowerBounds.size == 1 -> lowerBounds[0].asRuntimeType(typeVariableAccessor)
+                    upperBounds[0] == TypeName.OBJECT -> throw AssertionError()
+                    else -> upperBounds[0].asRuntimeType(typeVariableAccessor)
+                }
                 is TypeVariableName -> typeVariableAccessor(this)
                 else -> CodeBlock.of("\$T.class", this)
             }

--- a/compiler/src/main/kotlin/se/ansman/kotshi/Property.kt
+++ b/compiler/src/main/kotlin/se/ansman/kotshi/Property.kt
@@ -53,6 +53,10 @@ class Property(
     init {
         require(getter != null || field != null)
 
+        if (type.containsUnboundWildcards()) {
+            throw ProcessingError("Properties cannot contain wildcard types", parameter)
+        }
+
         defaultValueProvider = if (defaultValueQualifier != null || parameter.hasAnnotation<JsonDefaultValue>()) {
             if (adapterKey.isGeneric) {
                 throw ProcessingError("You cannot use default values on a generic type", parameter)

--- a/compiler/src/main/kotlin/se/ansman/kotshi/TypeNameExt.kt
+++ b/compiler/src/main/kotlin/se/ansman/kotshi/TypeNameExt.kt
@@ -2,6 +2,8 @@ package se.ansman.kotshi
 
 import com.squareup.javapoet.ParameterizedTypeName
 import com.squareup.javapoet.TypeName
+import com.squareup.javapoet.TypeVariableName
+import com.squareup.javapoet.WildcardTypeName
 
 val TYPE_NAME_STRING = TypeName.get(String::class.java)
 
@@ -26,3 +28,15 @@ val TypeName.rawType: TypeName
         is ParameterizedTypeName -> this.rawType
         else -> this
     }
+
+fun TypeName.containsUnboundWildcards(): Boolean =
+        when (this) {
+            is ParameterizedTypeName -> typeArguments.any { it.containsUnboundWildcards() }
+            is WildcardTypeName -> when {
+                lowerBounds.size == 1 -> lowerBounds[0].containsUnboundWildcards()
+                upperBounds[0] == TypeName.OBJECT -> true
+                else -> upperBounds[0].containsUnboundWildcards()
+            }
+            is TypeVariableName -> false
+            else -> false
+        }

--- a/tests/src/main/kotlin/se/ansman/kotshi/Wildcards.kt
+++ b/tests/src/main/kotlin/se/ansman/kotshi/Wildcards.kt
@@ -1,0 +1,4 @@
+package se.ansman.kotshi
+
+@JsonSerializable
+data class Wildcards(var list: List<CharSequence>)

